### PR TITLE
[CPDLP-2859] TS1 - Statements endpoint - happy path (show/index)

### PR DESCRIPTION
**Pull request by mooktakim**
### Context

Ticket: https://dfedigital.atlassian.net/browse/CPDLP-2859

### Changes proposed in this pull request

* `API::V3::StatementsController` updated with `index` and `show` actions
* Tests added

### CODE REVIEW

File: app/controllers/api/v3/statements_controller.rb [hunk 1]
```diff
@@ -1,8 +1,21 @@
...
   module V3
     class StatementsController < BaseController
-      def show = head(:method_not_allowed)
-      def index = head(:ok)
+      def index
+        render json: StatementSerializer.render(statements_query.statements, root: "data")
+      end
+
+      def show
+        render json: StatementSerializer.render(statements_query.statement(id: params[:id]), root: "data")
```
Comment (peteryates) [inline app/controllers/api/v3/statements_controller.rb:9]: Having to add `root: "data"` everywhere isn't the prettiest 😞 

[Doesn't look like there's any way around it either](https://github.com/procore-oss/blueprinter/issues/253).

File: spec/requests/api/v3/statements_spec.rb [hunk 2]
```diff
@@ -1,13 +1,36 @@
...
-        expect(response.status).to eq 200
-        expect(response.headers["Content-Type"]).to include("application/json")
+        it "returns 2 statements" do
+          api_get("/api/v3/statements")
+
+          expect(response.status).to eq 200
+          expect(response.content_type).to eql("application/json")
+          expect(parsed_response["data"].size).to eq(2)
+          expect(parsed_response["data"][0]["id"]).to eq(statement1.id)
+          expect(parsed_response["data"][1]["id"]).to eq(statement2.id)
```
Comment (peteryates) [inline spec/requests/api/v3/statements_spec.rb:23]: I'm not sure we should be testing the actual response here - shouldn't we trust our query object and serialiser? Otherwise we end up with request specs overlapping with both serialiser and query object ones.

The controller's responsibility here is just to respond to a request appropriately.

File: app/controllers/api/v3/statements_controller.rb (hunk 1 above)
Comment (cwrw) [reply app/controllers/api/v3/statements_controller.rb:9]: Maybe we can add a wrapper for this method in the base controller, and pass in the serialiser/params to avoid calling this everywhere?

Agree though its not great, and as @mooktakim mentioned in another conversation keep our options open to using other serialisers if we feel it might be a problem

Comment (cwrw) [reply app/controllers/api/v3/statements_controller.rb:9]: could we also add strong params, something similar to:
```
def statement_params
  params.permit(:id)
end
```

File: spec/requests/api/v3/statements_spec.rb (hunk 2 above)
Comment (mooktakim) [reply spec/requests/api/v3/statements_spec.rb:23]: I think its important to make sure the result is as we expect. In this case, we want to make sure the correct statements are returned, so only checked for the `id`'s. I shouldn't check all values as that's tested in the serializer unit tests.

Comment (peteryates) [reply spec/requests/api/v3/statements_spec.rb:23]: We end up in a situation where changes to the serializers or query objects mean we need to update API specs though

File: app/controllers/api/v3/statements_controller.rb (hunk 1 above)
Comment (peteryates) [reply app/controllers/api/v3/statements_controller.rb:9]: Also if we stick with BluePrinter and extend the query objects so they're modular and chainable, we could put th serialization there make this controller a little less involved, something more like:

```ruby
render(json: Statements::Query.new.for_lead_provider(current_lead_provider).serialize(:v3))
```

Comment (mooktakim) [reply app/controllers/api/v3/statements_controller.rb]: Update code with strong params 👍

Comment (mooktakim) [reply app/controllers/api/v3/statements_controller.rb]: `root: "data"` definitely not nice. I'm actually leaning towards going back to `jsonapi-serializer` 😂

File: spec/requests/api/v3/statements_spec.rb (hunk 2 above)
Comment (mooktakim) [reply spec/requests/api/v3/statements_spec.rb:23]: We have to check whether we return the correct objects at least, otherwise its not really end to end.

Comment (cwrw) [reply spec/requests/api/v3/statements_spec.rb:23]: I was having a think about this, we need to exercise some form of reliable end to end testing, without going to deep into retesting the unit tests covered in serialisers/queries. I think checking size/ids is a good balance. If we stick to the following structure:
- Unit tests for serialisers
- Unit tests for queries and all edge cases
- Tests for ensuring schema is met for swagger
- Request specs which check statuses are ok, size and ids are ok and are very light happy path scenarios as in this PR it would be enough without delving into the fields returned and should cover us, otherwise I feel like we'll miss something if it does break

File: app/controllers/api/v3/statements_controller.rb (hunk 1 above)
Comment (mooktakim) [reply app/controllers/api/v3/statements_controller.rb]: I tried to DRY up the code a bit with `to_json` method

File: app/controllers/api/v3/statements_controller.rb [hunk 3]
```diff
@@ -1,8 +1,29 @@
...
+
+    private
+
+      def statements_query
+        ::Statements::Query.new(
+          lead_provider: current_lead_provider,
+        )
+      end
+
+      def statement_params
```
[the remark this answers is not in the document]
Comment (cwrw) [reply app/controllers/api/v3/statements_controller.rb:20]: Although no mass assignments is happening, I think it's good practice to use and we'll be adding more params soon like filtering etc. so its safe so start early

**Review by cwrw — COMMENTED** [review]
looks great @mooktakim! only last comment would be around the order of params appearing in the API responses, testing locally they are ordered alphabetically not sure if you're also seeing this? any way we can preserve the expected order instead?

**Comment by mooktakim** [comment]
> looks great @mooktakim! only last comment would be around the order of params appearing in the API responses, testing locally they are ordered alphabetically not sure if you're also seeing this? any way we can preserve the expected order instead?

I was able to add attributes ordering based on definition 👍

**Review by cwrw — APPROVED** [review]
thanks mook 👍
